lib/color_spinor_wuppertal.cu

```diff
@@ -0,0 +1,296 @@
+#include <transfer.h>
+#include <quda_internal.h>
+#include <quda_matrix.h>
+#include <gauge_field.h>
+#include <gauge_field_order.h>
+#include <index_helper.cuh>
+#include <color_spinor.h>
+#include <color_spinor_field.h>
+#include <color_spinor_field_order.h>
+#include <tune_quda.h>
+
+namespace quda {
+
+  template <typename Float, int Ns, int Nc, QudaReconstructType gRecon>
+  struct WuppertalSmearingArg {
+    typedef typename colorspinor_mapper<Float,Ns,Nc>::type F;
+    typedef typename gauge_mapper<Float,gRecon>::type G;
+
+    F out;                // output vector field
+    const F in;           // input vector field
+    const G U;            // the gauge field
+    const Float A;        // A parameter
+    const Float B;        // B parameter
+    const int parity;     // only use this for single parity fields
+    const int nParity;    // number of parities we're working on
+    const int nFace;      // hard code to 1 for now
+    const int dim[5];     // full lattice dimensions
+    const int commDim[4]; // whether a given dimension is partitioned or not
+    const int volumeCB;   // checkerboarded volume
+
+    WuppertalSmearingArg(ColorSpinorField &out, const ColorSpinorField &in, int parity, const GaugeField &U,
+                       Float A, Float B)
+      : out(out), in(in), U(U), A(A), B(B), parity(parity), nParity(in.SiteSubset()), nFace(1),
+        dim{ (3-nParity) * in.X(0), in.X(1), in.X(2), in.X(3), 1 },
+      commDim{comm_dim_partitioned(0), comm_dim_partitioned(1), comm_dim_partitioned(2), comm_dim_partitioned(3)},
+      volumeCB(in.VolumeCB())
+    {
+      if (in.FieldOrder() != QUDA_FLOAT2_FIELD_ORDER || !U.isNative())
+        errorQuda("Unsupported field order colorspinor=%d gauge=%d combination\n", in.FieldOrder(), U.FieldOrder());
+    }
+  };
+
+  /**
+     Computes out = sum_mu U_mu(x)in(x+d) + U^\dagger_mu(x-d)in(x-d)
+     @param[out] out The out result field
+     @param[in] U The gauge field
+     @param[in] in The input field
+     @param[in] x_cb The checkerboarded site index
+     @param[in] parity The site parity
+  */
+  template <typename Float, int Nc, typename Vector, typename Arg>
+  __device__ __host__ inline void computeNeighborSum(Vector &out, Arg &arg, int x_cb, int parity) {
+
+    typedef Matrix<complex<Float>,Nc> Link;
+    const int their_spinor_parity = (arg.nParity == 2) ? 1-parity : 0;
+
+    int coord[5];
+    getCoords(coord, x_cb, arg.dim, parity);
+    coord[4] = 0;
+
+#pragma unroll
+    for (int dir=0; dir<3; dir++) { // loop over spatial directions
```

Comment:
Just a thought here. Given that the smearing operator is constant, is there some way that one can gather all the links first, then loop over the SU(3) mat-vec operations? There would need to be a synchronization guard in place so that `out` vectors are not written before they are gathered as `in` vectors in the same smearing step.

This would reduce memory traffic considerably. For a typical O(100) smearing step operation.